### PR TITLE
make: add help target to root Makefile for printing info about availble targets

### PR DESCRIPTION
**Pull request by fristonio**
Adds a new help target to the root Makefile that can be used to print information about the available targets from the Makefile. This can be pretty useful given the large number of make targets we have. 

Sample: 

image

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -144,7 +150,7 @@ tests-privileged:
 	done
 	$(MAKE) generate-cov
 
-start-kvstores:
+start-kvstores: ## Start running KVStores required for running Cilium unit-tests. More specifically this will run etcd and consul containers.
```
Comment (christarazi) [inline Makefile]: Nit: for consistency
```suggestion
start-kvstores: ## Start running kvstores required for running Cilium unit-tests. More specifically this will run etcd and consul containers.
```

File: Makefile [hunk 2]
```diff
@@ -215,17 +221,17 @@ endif
 	$(MAKE) generate-cov
 	$(MAKE) stop-kvstores
 
-bench: start-kvstores
+bench: start-kvstores ## Run benchmarks for cilium unit-tests in the repository.
```
Comment (christarazi) [inline Makefile]: Nit: 
```suggestion
bench: start-kvstores ## Run benchmarks for Cilium unit-tests in the repository.
```

File: Makefile [hunk 3]
```diff
@@ -234,46 +240,47 @@ bench-privileged:
 		|| exit 1; \
 	done
 
-clean-tags:
+clean-tags: ## Remove all the cscope tags file from the repository.
```
Comment (christarazi) [inline Makefile]:
```suggestion
clean-tags: ## Remove all the tags files from the repository.
```

File: Makefile [hunk 4]
```diff
@@ -234,46 +240,47 @@ bench-privileged:
...
 	@echo $(GOLANG_SRCFILES) $(BPF_SRCFILES) | sed 's/ /\n/g' | sort > cscope.files
 
-tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES) cscope.files
+tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES) cscope.files ## Generate tags for golang and bpf source files.
 	@ctags $(GOLANG_SRCFILES) $(BPF_SRCFILES)
 	cscope -R -b -q
 
-clean-container:
-	-$(QUIET) for i in $(SUBDIRS); do $(MAKE) $(SUBMAKEOPTS) -C $$i clean; done
+clean-container: ## Perform cleanup for components required in cilium-agent container.
```
Comment (christarazi) [inline Makefile]:
```suggestion
clean-container: ## Perform `make clean` for each component required in cilium-agent container.
```

File: Makefile [hunk 5]
```diff
@@ -234,46 +240,47 @@ bench-privileged:
...
+	-$(QUIET) for i in $(SUBDIRS_CILIUM_CONTAINER); do $(MAKE) $(SUBMAKEOPTS) -C $$i clean; done
 
-clean: clean-container
+clean: ## Perform overall cleanup for Cilium.
+	-$(QUIET) for i in $(SUBDIRS); do $(MAKE) $(SUBMAKEOPTS) -C $$i clean; done
 	-$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C ./contrib/packaging/deb clean
 	-$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C ./contrib/packaging/rpm clean
 
-veryclean:
+ctrclean: ## Perform complete cleanup for container engine images(including build cache).
```
Comment (christarazi) [inline Makefile]: Why was this renamed?

File: Makefile [hunk 6]
```diff
@@ -234,46 +240,47 @@ bench-privileged:
...
 	-$(QUIET) $(CONTAINER_ENGINE) builder prune -af
 
-install-bpf:
+install-bpf: ## Copies over the BPF source files from bpf/ to /var/lib/cilium/bpf/
 	$(QUIET)$(INSTALL) -m 0750 -d $(DESTDIR)$(LOCALSTATEDIR)/lib/cilium
 	-rm -rf $(DESTDIR)$(LOCALSTATEDIR)/lib/cilium/bpf/*
 	$(foreach bpfsrc,$(BPF_SRCFILES), $(INSTALL) -D -m 0644 $(bpfsrc) $(DESTDIR)$(LOCALSTATEDIR)/lib/cilium/$(bpfsrc);)
 
-install: install-bpf
+install: install-bpf ## Performs install for all the cilium sub components(daemon, operator, relay etc.)
```
Comment (christarazi) [inline Makefile]:
```suggestion
install: install-bpf ## Performs install for all the Cilium sub components (daemon, operator, relay etc.)
```

File: Makefile [hunk 7]
```diff
@@ -234,46 +240,47 @@ bench-privileged:
...
 	$(QUIET)$(INSTALL) -m 0755 -d $(DESTDIR)$(BINDIR)
 	for i in $(SUBDIRS); do $(MAKE) $(SUBMAKEOPTS) -C $$i install; done
 
-install-container: install-bpf
+install-container: install-bpf ## Performs install for all components required with cilium-agent.
```
Comment (christarazi) [inline Makefile]:
```suggestion
install-container: install-bpf ## Performs install for all components required for cilium-agent container.
```

Comment (christarazi) [reply Makefile]: There are a few cases of these below

File: Makefile [hunk 8]
```diff
@@ -282,11 +289,23 @@ install-bash-completion:
...
+k8s-tests:
+	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C tests k8s-tests
+
 -include Makefile.docker
 
+##@ API targets
 CRD_OPTIONS ?= "crd:crdVersions=v1"
-# Generate manifests e.g. CRD, RBAC etc.
-manifests:
+manifests: ## Generate manifests e.g. CRD, RBAC etc.
```
Comment (christarazi) [inline Makefile]: Now that we are labeling this "officially", let's make it be slightly more specific:

```suggestion
manifests: ## Generate K8s manifests e.g. CRD, RBAC etc.
```

File: Makefile [hunk 9]
```diff
@@ -299,19 +318,7 @@ manifests:
...
-	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C ./contrib/packaging/rpm
-
-runtime-tests:
-	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C tests runtime-tests
-
-k8s-tests:
-	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C tests k8s-tests
-
-generate-api: api/v1/openapi.yaml
+generate-api: api/v1/openapi.yaml ## Generate cilium-agent client, model and server from openapi spec.
```
Comment (christarazi) [inline Makefile]:
```suggestion
generate-api: api/v1/openapi.yaml ## Generate cilium-agent client, model and server code from openapi spec.
```

Comment (christarazi) [reply Makefile]: A few more instances below

**Review by christarazi — CHANGES_REQUESTED** [review]
:rocket: This is great, thanks for the PR! Some nits below

**Comment by aditighag** [comment]
Super cool! Won't have to reference multiple guides to find a build target now. Could you also add `make debug` to the help output? Something along these lines - 

`debug Builds Cilium by disabling inlining and compiler optimizations. Useful for debugging.`

**Comment by christarazi** [comment]
@aditighag It would also be good to mention that it doesn't strip the binaries of debug symbols as well.

File: Makefile (hunk 5 above)
Comment (fristonio) [reply Makefile]: Had this local change from before, will update the target name.

**Review by twpayne — COMMENTED** [review]
Very cool!

File: Makefile [hunk 10]
```diff
@@ -122,17 +128,17 @@ define generate_k8s_protobuf
...
 	for i in $(SUBDIRS_CILIUM_CONTAINER); do $(MAKE) $(SUBMAKEOPTS) -C $$i all; done
 
-$(SUBDIRS): force
+$(SUBDIRS): force ## Execute default make target(make all) for the provided subdirectory.
 	@ $(MAKE) $(SUBMAKEOPTS) -C $@ all
 
 PRIV_TEST_PKGS_EVAL := $(shell for pkg in $(TESTPKGS); do echo $$pkg; done | xargs grep --include='*.go' -ril '+build [^!]*privileged_tests' | xargs dirname | sort | uniq)
 PRIV_TEST_PKGS ?= $(PRIV_TEST_PKGS_EVAL)
-tests-privileged: GO_TAGS_FLAGS+=privileged_tests
+tests-privileged: GO_TAGS_FLAGS+=privileged_tests ## Run unit-tests for cilium that requires elevated previliges.
```
Comment (twpayne) [inline Makefile]: typo: s/previliges/privileges/

File: Makefile [hunk 11]
```diff
@@ -122,17 +128,17 @@ define generate_k8s_protobuf
 	    --go-header-file "$(PWD)/hack/custom-boilerplate.go.txt"
 endef
 
-build: $(SUBDIRS)
+build: $(SUBDIRS) ## Builds all the components for cilium by executing make in the respective sub directories.
```
Comment (christarazi) [inline Makefile]: A few instances below as well where the capitalization of "cilium" is inconsistent

**Review by christarazi — APPROVED** [review]
LGTM, a non-blocking nit below.

File: Makefile (hunk 11 above)
Comment (fristonio) [reply Makefile]: Thanks for pointing this out, I have updated the instances.

File: Makefile [hunk 12]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET) $(MAKE) -C images update-golang-image
 	@echo "Updated go version in image Dockerfiles to $(GO_IMAGE_VERSION)"
 
-dev-doctor:
+dev-doctor: ## Run Cilium dev-doctor to validate local development environment.
 	$(QUIET)$(GO) version 2>/dev/null || ( echo "go not found, see https://golang.org/doc/install" ; false )
 	$(QUIET)$(GO) run ./tools/dev-doctor
 
-.PHONY: clean clean-container dev-doctor force generate-api generate-health-api generate-operator-api generate-hubble-api install licenses-all veryclean
+help: Makefile ## Display help for the Makefile, from https://suva.sh/posts/well-documented-makefiles/
```
Comment (sayboras) [inline Makefile]: nit: this link is no longer valid, maybe https://www.thapaliya.com/en/writings/well-documented-makefiles/

File: Makefile [hunk 13]
```diff
@@ -173,29 +179,29 @@ ifeq ($(SKIP_KVSTORES),"false")
...
-stop-kvstores:
+stop-kvstores: ## Forcefully removes running kvstore components for Cilium unit-testing(etcd and consul containers).
 ifeq ($(SKIP_KVSTORES),"false")
 	$(QUIET)$(CONTAINER_ENGINE) rm -f "cilium-etcd-test-container"
 	$(QUIET)$(CONTAINER_ENGINE) rm -f "cilium-consul-test-container"
 	$(QUIET)rm -rf /tmp/cilium-consul-certs
 endif
 
-tests: force
+tests: force ## Run all unit tests for Cilium
```
Comment (sayboras) [inline Makefile]: nit: mention _force_ to differentiate with unit-tests

**Review by sayboras — APPROVED** [review]
few nits as below, and maybe cilium -> Cilium. 

I like your changes with docker related targets :100:, now i don't need to check Makefile code to find correct target. It's super hard to to do grouping for Makefile.

LGTM :100:

File: Makefile (hunk 13 above)
Comment (fristonio) [reply Makefile]: I am not sure if I understand the difference between the targets correctly. The `force` forces rebuild of `tests` target but inherently both `tests` and `unit-tests` seems the same to me, rebuilt every time.
Is there something I am missing here?

File: Makefile (hunk 12 above)
Comment (fristonio) [reply Makefile]: Good Catch, will update the link.

File: Makefile [hunk 14]
```diff
@@ -234,46 +240,47 @@ bench-privileged:
...
+clean-tags: ## Remove all the tags files from the repository.
 	@$(ECHO_CLEAN) tags
 	@-rm -f cscope.out cscope.in.out cscope.po.out cscope.files tags
 
-cscope.files: $(GOLANG_SRCFILES) $(BPF_SRCFILES)
+cscope.files: $(GOLANG_SRCFILES) $(BPF_SRCFILES) ## Generate cscope.files with the list of all files to generate ctags for.
 	@echo $(GOLANG_SRCFILES) $(BPF_SRCFILES) | sed 's/ /\n/g' | sort > cscope.files
 
-tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES) cscope.files
+tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES) cscope.files ## Generate tags for golang and bpf source files.
```
Comment (tklauser) [inline Makefile]: Nit:

```suggestion
tags: $(GOLANG_SRCFILES) $(BPF_SRCFILES) cscope.files ## Generate tags for Go and BPF source files.
```

File: Makefile [hunk 15]
```diff
@@ -362,10 +369,10 @@ generate-operator-api: api/v1/operator/openapi.yaml
 	@# sort goimports automatically
 	-$(QUIET) find api/v1/operator/ -type f -name "*.go" -print | PATH="$(PWD)/tools:$(PATH)" xargs goimports -w
 
-generate-hubble-api: api/v1/flow/flow.proto api/v1/peer/peer.proto api/v1/observer/observer.proto api/v1/relay/relay.proto
+generate-hubble-api: api/v1/flow/flow.proto api/v1/peer/peer.proto api/v1/observer/observer.proto api/v1/relay/relay.proto ## Generate hubble proto golang sources.
```
Comment (tklauser) [inline Makefile]:
```suggestion
generate-hubble-api: api/v1/flow/flow.proto api/v1/peer/peer.proto api/v1/observer/observer.proto api/v1/relay/relay.proto ## Generate hubble proto Go sources.
```

File: Makefile [hunk 16]
```diff
@@ -362,10 +369,10 @@ generate-operator-api: api/v1/operator/openapi.yaml
...
 	$(QUIET) $(MAKE) $(SUBMAKEOPTS) -C api/v1
 
-generate-k8s-api:
+generate-k8s-api: ## Generate Cilium k8s API client, deepcopy and deepequal golang sources.
```
Comment (tklauser) [inline Makefile]:
```suggestion
generate-k8s-api: ## Generate Cilium k8s API client, deepcopy and deepequal Go sources.
```

File: Makefile [hunk 17]
```diff
@@ -458,10 +466,10 @@ release:
 	git tag v$(VERSION)
 	git archive --format tar $(BRANCH) | gzip > ../cilium_$(VERSION).orig.tar.gz
 
-gofmt:
+gofmt: ## Run gofmt on golang source files in the repository.
```
Comment (tklauser) [inline Makefile]:
```suggestion
gofmt: ## Run gofmt on Go source files in the repository.
```

File: Makefile [hunk 18]
```diff
@@ -458,10 +466,10 @@ release:
...
 	$(QUIET)for pkg in $(GOFILES); do $(GO) fmt $$pkg; done
 
-govet:
+govet: ## Run govet on golang source files in the repository.
```
Comment (tklauser) [inline Makefile]:
```suggestion
govet: ## Run govet on Go source files in the repository.
```

File: Makefile [hunk 19]
```diff
@@ -482,15 +490,15 @@ govet:
     ./test/k8sT/... \
     ./tools/...
 
-lint:
+lint: ## Run golangci-lint.
 	@$(ECHO_CHECK) golangci-lint
 	$(QUIET) golangci-lint run
 
-logging-subsys-field:
+logging-subsys-field: ## Validate logrus subsystem field for logs in golang source code.
```
Comment (tklauser) [inline Makefile]:
```suggestion
logging-subsys-field: ## Validate logrus subsystem field for logs in Go source code.
```

File: Makefile [hunk 20]
```diff
@@ -527,92 +535,96 @@ endif
...
-pprof-help:
-	@echo "Available pprof targets:"
-	@echo "  pprof-heap"
-	@echo "  pprof-profile"
-	@echo "  pprof-block"
-	@echo "  pprof-trace-5s"
-	@echo "  pprof-mutex"
-
-pprof-heap:
+pprof-heap: ## Get golang heap profile.
```
Comment (tklauser) [inline Makefile]:
```suggestion
pprof-heap: ## Get Go pprof heap profile.
```

File: Makefile [hunk 21]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET)$(GO) tool pprof http://localhost:6060/debug/pprof/heap
 
-pprof-profile:
+pprof-profile: ## Get golang pprof profile
```
Comment (tklauser) [inline Makefile]:
```suggestion
pprof-profile: ## Get Go pprof profile.
```

File: Makefile [hunk 22]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET)$(GO) tool pprof http://localhost:6060/debug/pprof/profile
 
-
-pprof-block:
+pprof-block: ## Get golang pprof block profile
```
Comment (tklauser) [inline Makefile]:
```suggestion
pprof-block: ## Get Go pprof block profile.
```

File: Makefile [hunk 23]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET)$(GO) tool pprof http://localhost:6060/debug/pprof/block
 
-pprof-trace-5s:
+pprof-trace-5s: ## Get golang pprof trace for a duration of 5 seconds.
```
Comment (tklauser) [inline Makefile]:
```suggestion
pprof-trace-5s: ## Get Go pprof trace for a duration of 5 seconds.
```

File: Makefile [hunk 24]
```diff
@@ -527,92 +535,96 @@ endif
...
 	curl http://localhost:6060/debug/pprof/trace?seconds=5
 
-pprof-mutex:
+pprof-mutex: ## Get golang pprof mutex profile.
```
Comment (tklauser) [inline Makefile]:
```suggestion
pprof-mutex: ## Get Go pprof mutex profile.
```

File: Makefile [hunk 25]
```diff
@@ -527,92 +535,96 @@ endif
...
 	mkdir -p man
 	cilium cmdman -d man
 
-install-manpages:
+install-manpages: ## Install manpages the Cilium CLI.
 	cp man/* /usr/local/share/man/man1/
 	mandb
 
-postcheck: build
+postcheck: build ## Run Cilium build postcheck(update-cmdref, build documentation etc.)
```
Comment (tklauser) [inline Makefile]:
```suggestion
postcheck: build ## Run Cilium build postcheck (update-cmdref, build documentation etc.)
```

File: Makefile [hunk 26]
```diff
@@ -527,92 +535,96 @@ endif
...
-minikube:
+minikube: ## Setup a minikube with Cilium installed.
 	$(QUIET) contrib/scripts/minikube.sh
 
-licenses-all:
+licenses-all: ## Generate file with all the License from dependencies.
 	@$(GO) run ./tools/licensegen > LICENSE.all || ( rm -f LICENSE.all ; false )
 
-update-go-version: update-dev-doctor-go-version update-gh-actions-go-version update-travis-go-version update-test-go-version update-images-go-version
+update-go-version: update-dev-doctor-go-version update-gh-actions-go-version update-travis-go-version update-test-go-version update-images-go-version ## Update go version for all the components(travis, dev-doctor, gh-actions etc.).
```
Comment (tklauser) [inline Makefile]:
```suggestion
update-go-version: update-dev-doctor-go-version update-gh-actions-go-version update-travis-go-version update-test-go-version update-images-go-version ## Update Go version for all the components (travis, dev-doctor, gh-actions etc.).
```

File: Makefile [hunk 27]
```diff
@@ -527,92 +535,96 @@ endif
...
 
-update-dev-doctor-go-version:
+update-dev-doctor-go-version: ## Update dev-doctor golang version.
```
Comment (tklauser) [inline Makefile]:
```suggestion
update-dev-doctor-go-version: ## Update dev-doctor Go version.
```

File: Makefile [hunk 28]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET) sed -i 's/^const minGoVersionStr = ".*"/const minGoVersionStr = "$(GO_MAJOR_AND_MINOR_VERSION)"/' tools/dev-doctor/config.go
 	@echo "Updated go version in tools/dev-doctor to $(GO_MAJOR_AND_MINOR_VERSION)"
 
-update-gh-actions-go-version:
+update-gh-actions-go-version: ## Update GitHub action golang version.
```
Comment (tklauser) [inline Makefile]:
```suggestion
update-gh-actions-go-version: ## Update Go version in GitHub action config.
```

File: Makefile [hunk 29]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET) for fl in $(shell find .github/workflows -name "*.yaml" -print) ; do sed -i 's/go-version: .*/go-version: $(GO_IMAGE_VERSION)/g' $$fl ; done
 	@echo "Updated go version in GitHub Actions to $(GO_IMAGE_VERSION)"
 
-update-travis-go-version:
+update-travis-go-version: ## Update golang version in TravisCI config.
```
Comment (tklauser) [inline Makefile]:
```suggestion
update-travis-go-version: ## Update Go version in TravisCI config.
```

File: Makefile [hunk 30]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET) sed -i 's/go: ".*/go: "$(GO_VERSION)"/g' .travis.yml
 	@echo "Updated go version in .travis.yml to $(GO_VERSION)"
 
-update-test-go-version:
+update-test-go-version: ## Update golang version in test scripts.
```
Comment (tklauser) [inline Makefile]:
```suggestion
update-test-go-version: ## Update Go version in test scripts.
```

File: Makefile [hunk 31]
```diff
@@ -527,92 +535,96 @@ endif
...
 	$(QUIET) sed -i 's/GO_VERSION=.*/GO_VERSION="$(GO_VERSION)"/g' test/kubernetes-test.sh
 	$(QUIET) sed -i 's/GOLANG_VERSION=.*/GOLANG_VERSION="$(GO_VERSION)"/g' test/packet/scripts/install.sh
 	@echo "Updated go version in test scripts to $(GO_VERSION)"
 
-update-images-go-version:
+update-images-go-version: ## Update golang version in Dockerfiles.
```
Comment (tklauser) [inline Makefile]:
```suggestion
update-images-go-version: ## Update Go version in Dockerfiles.
```

**Review by tklauser — COMMENTED** [review]
Nice work, this is really `help`ful 🚀

Some spelling consistency suggestions inline, mainly `s/golang/Go` as per https://golang.org/doc/faq#go_or_golang

File: Makefile (hunk 13 above)
Comment (sayboras) [reply Makefile]: oh right, in that case, we might consider to keep one only. I personally didn't know that `tests` is there, I used unit-tests mainly.

**Review by tklauser — APPROVED** [review]
:rocket:

**Review by twpayne — APPROVED** [review]
🍬
